### DISCUSSION TITLE
[Enhancement] Performance Profiler

### DISCUSSION THREAD

**Discussion by jrc14** [dotnet/maui#154] [General] [answered]
## Summary
Make the performance profiling tools available in the Community toolkit - or at least available in something cheaper than an Enterprise subscription to VS.

## Intended Use Case
Developing solid production apps in Android is really difficult without tools for profiling (performance and memory allocation) , but an enterprise subscription is just too expensive.

**Comment by awulkan** [comment]
For comparison, Flutter offers great performance profiling tools for free. If MAUI wants to be competitive I think this is a must.

**Comment by keozx** [comment]
Even making the current xamarin profiler work again would do, but it seems since 2019 there are no resources allocated to keep working on it, maybe open source this as well.
In fact, .NET 6 to "any device" profiling should be a priority since now the main selling point will be one dot net, everywhere, right? not only web and desktop

**Comment by Redth** [comment, marked as answer]
This is something we are looking at adding in the future.

For now, there is some support for capturing profile information and you can use tools like speedscope, or load the captured traces into the VS tools directly.

https://github.com/dotnet/maui/wiki/Profiling-.NET-MAUI-Apps

**Reply by jonathanpeppers** [reply]
More docs on this here: https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/tracing.md

**Reply by Dreamescaper** [reply]
@Redth @jonathanpeppers 
Is it possible to use dotnet-trace to profile memory usage (allocations, GCs)? I've tried to run it with `--profile gc-verbose`, but it does not produce any results.
Thanks!

**Reply by jonathanpeppers** [reply]
@lateralusX can probably comment on this ^^

I think you can get raw events, but they might be hard to sift through. You would need to open the `.nettrace` file in PerfView.

**Reply by lateralusX** [reply]
Current EventPipe GC events are CoreCLR specific and not yet supported on Mono. I did some 1:1 mapping of old Mono profile events over to nettrace file format when implementing diagnostics and EventPipe support on Mono. As part of those events, there will also be GC events that can be used to look at GC allocations. I have described that here https://github.com/dotnet/runtime/blob/main/docs/design/mono/diagnostics-tracing.md#collect-gc-dumps-on-monovm, and I made a couple of samples showing how to parse nettrace files using TraceEvent library here, https://github.com/lateralusX/diagnostics-nettrace-samples.

**Comment by Xyncgas** [comment]
how about just use the force to sense complexity of your codes
